AWS VpcPeeringConnections parser error

If the VpcPeeringConnections-<region>.json file has a stanza for Acceptor and Requestor VPC info, which is missing the CidrBlock entry

   "CidrBlock": "10.194.0.0/16", 
    "CidrBlockSet": [
     {
      "CidrBlock": "10.194.0.0/16"
     }

then it fails. This scenario could arise if the VpcPeeringConnection was deleted (as shown below):

{
   "AccepterVpcInfo": {
    "OwnerId": "118292266645", 
    "VpcId": "vpc-6ee90d17"
   }, 
   "RequesterVpcInfo": {
    "OwnerId": "118292266645", 
    "VpcId": "vpc-56c5212f"
   }, 
   "Status": {
    "Code": "deleted", 
    "Message": "Deleted by 118292266645"
   },

The error message seen is:

org.batfish.common.BatfishException: Problems parsing JSON in /home/samir/git/barclays-demo/containers/LHR/testrigs/master_2017-12-06-22:20:17_655a652/testrig/aw
s_configs/VpcPeeringConnections-us-west-2.json
	at org.batfish.main.Batfish.parseAwsVpcConfigurations(Batfish.java:2793)
	at org.batfish.main.Batfish.serializeAwsVpcConfigs(Batfish.java:3910)
	at org.batfish.main.Batfish.serializeVendorConfigs(Batfish.java:4202)
	at org.batfish.main.Batfish.run(Batfish.java:3845)
	at org.batfish.main.Driver$1.run(Driver.java:356)
Caused by: org.codehaus.jettison.json.JSONException: JSONObject["CidrBlock"] not found.
	at org.codehaus.jettison.json.JSONObject.get(JSONObject.java:407)
	at org.codehaus.jettison.json.JSONObject.getString(JSONObject.java:543)
	at org.batfish.representation.aws_vpcs.VpcPeeringConnection.<init>(VpcPeeringConnection.java:28)
	at org.batfish.representation.aws_vpcs.AwsVpcConfiguration.addConfigElement(AwsVpcConfiguration.java:139)
	at org.batfish.representation.aws_vpcs.AwsVpcConfiguration.addConfigElement(AwsVpcConfiguration.java:78)
	at org.batfish.main.Batfish.parseAwsVpcConfigurations(Batfish.java:2791)
	... 4 more
{
  "answerElements" : [
    {
      "class" : "org.batfish.common.BatfishException$BatfishStackTrace",
      "answer" : [
        "org.batfish.common.BatfishException: Batfish job failed",
        "   at org.batfish.main.Driver$1.run(Driver.java:384)",
        "Caused by: org.batfish.common.BatfishException: Problems parsing JSON in /home/samir/git/barclays-demo/containers/LHR/testrigs/master_2017-12-06-22:20:1
7_655a652/testrig/aws_configs/VpcPeeringConnections-us-west-2.json",
        "   at org.batfish.main.Batfish.parseAwsVpcConfigurations(Batfish.java:2793)",
        "   at org.batfish.main.Batfish.serializeAwsVpcConfigs(Batfish.java:3910)",
        "   at org.batfish.main.Batfish.serializeVendorConfigs(Batfish.java:4202)",
        "   at org.batfish.main.Batfish.run(Batfish.java:3845)",
        "   at org.batfish.main.Driver$1.run(Driver.java:356)",
        "Caused by: org.codehaus.jettison.json.JSONException: JSONObject[\"CidrBlock\"] not found.",
        "   at org.codehaus.jettison.json.JSONObject.get(JSONObject.java:407)",
        "   at org.codehaus.jettison.json.JSONObject.getString(JSONObject.java:543)",
        "   at org.batfish.representation.aws_vpcs.VpcPeeringConnection.<init>(VpcPeeringConnection.java:28)",
        "   at org.batfish.representation.aws_vpcs.AwsVpcConfiguration.addConfigElement(AwsVpcConfiguration.java:139)",
        "   at org.batfish.representation.aws_vpcs.AwsVpcConfiguration.addConfigElement(AwsVpcConfiguration.java:78)",
        "   at org.batfish.main.Batfish.parseAwsVpcConfigurations(Batfish.java:2791)",
        "   ... 4 more"
      ]
    }
  ],
  "status" : "FAILURE",
  "summary" : {
    "numFailed" : 0,
    "numPassed" : 0,
    "numResults" : 0
  }
}

--- a/projects/batfish/src/main/java/org/batfish/representation/aws_vpcs/AwsVpcConfiguration.java
+++ b/projects/batfish/src/main/java/org/batfish/representation/aws_vpcs/AwsVpcConfiguration.java
@@ -136,8 +136,14 @@ public class AwsVpcConfiguration implements Serializable, GenericConfigObject {
         _vpcs.put(vpc.getId(), vpc);
         break;
       case AwsVpcEntity.JSON_KEY_VPC_PEERING_CONNECTIONS:
-        VpcPeeringConnection vpcPeerConn = new VpcPeeringConnection(jsonObject, logger);
-        _vpcPeerings.put(vpcPeerConn.getId(), vpcPeerConn);
+        String code =
+            jsonObject
+                .getJSONObject(AwsVpcEntity.JSON_KEY_STATUS)
+                .getString(AwsVpcEntity.JSON_KEY_CODE);
+        if (!code.equals(AwsVpcEntity.STATUS_DELETED)) {
+          VpcPeeringConnection vpcPeerConn = new VpcPeeringConnection(jsonObject, logger);
+          _vpcPeerings.put(vpcPeerConn.getId(), vpcPeerConn);
+        }
         break;
       case AwsVpcEntity.JSON_KEY_VPN_CONNECTIONS:
         VpnConnection vpnConnection = new VpnConnection(jsonObject, logger);

--- a/projects/batfish/src/main/java/org/batfish/representation/aws_vpcs/AwsVpcEntity.java
+++ b/projects/batfish/src/main/java/org/batfish/representation/aws_vpcs/AwsVpcEntity.java
@@ -14,6 +14,7 @@ public interface AwsVpcEntity {
   String JSON_KEY_BGP_ASN = "BgpAsn";
   String JSON_KEY_CIDR_BLOCK = "CidrBlock";
   String JSON_KEY_CIDR_IP = "CidrIp";
+  String JSON_KEY_CODE = "Code";
   String JSON_KEY_CUSTOMER_GATEWAY_CONFIGURATION = "CustomerGatewayConfiguration";
   String JSON_KEY_CUSTOMER_GATEWAY_ID = "CustomerGatewayId";
   String JSON_KEY_CUSTOMER_GATEWAYS = "CustomerGateways";
@@ -84,6 +85,8 @@ public interface AwsVpcEntity {
   String JSON_KEY_VPN_CONNECTIONS = "VpnConnections";
   String JSON_KEY_VPN_GATEWAY_ID = "VpnGatewayId";
   String JSON_KEY_VPN_GATEWAYS = "VpnGateways";
+
+  String STATUS_DELETED = "deleted";
 
   String XML_KEY_ASN = "asn";
   String XML_KEY_AUTHENTICATION_PROTOCOL = "authentication_protocol";

--- a/projects/batfish/src/test/java/org/batfish/grammar/aws/AwsVpcConfigurationTest.java
+++ b/projects/batfish/src/test/java/org/batfish/grammar/aws/AwsVpcConfigurationTest.java
@@ -1,0 +1,58 @@
+package org.batfish.grammar.aws;
+
+import static org.hamcrest.Matchers.hasKey;
+import static org.hamcrest.Matchers.not;
+import static org.junit.Assert.assertThat;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSortedSet;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.SortedSet;
+import org.batfish.common.BfConsts;
+import org.batfish.main.Batfish;
+import org.batfish.main.BatfishTestUtils;
+import org.batfish.main.TestrigText;
+import org.batfish.representation.aws_vpcs.AwsVpcConfiguration;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.rules.TemporaryFolder;
+
+public class AwsVpcConfigurationTest {
+
+  private static String TESTRIG_PREFIX = "org/batfish/grammar/aws/testrigs/";
+
+  @Rule public TemporaryFolder _folder = new TemporaryFolder();
+
+  @Rule public ExpectedException _thrown = ExpectedException.none();
+
+  @Test
+  public void testVpcPeeringConnections() throws IOException {
+    String testrigResourcePrefix = TESTRIG_PREFIX + "vpc_peering_connections";
+    SortedSet<String> awsFilenames = ImmutableSortedSet.of("VpcPeeringConnections.json");
+
+    Batfish batfish =
+        BatfishTestUtils.getBatfishFromTestrigText(
+            TestrigText.builder().setAwsText(testrigResourcePrefix, awsFilenames).build(), _folder);
+    batfish.loadConfigurations();
+    Path awsVendorFile =
+        batfish
+            .getTestrigSettings()
+            .getSerializeVendorPath()
+            .resolve(BfConsts.RELPATH_AWS_VPC_CONFIGS_FILE);
+    AwsVpcConfiguration awsVpcConfiguration =
+        batfish
+            .deserializeObjects(
+                ImmutableMap.of(awsVendorFile, BfConsts.RELPATH_AWS_VPC_CONFIGS_FILE),
+                AwsVpcConfiguration.class)
+            .get(BfConsts.RELPATH_AWS_VPC_CONFIGS_FILE);
+
+    /*
+     * We should have an entry for the vpc peering connection with status code "active", but not
+     * for the one with status code "deleted".
+     */
+    assertThat(awsVpcConfiguration.getVpcPeeringConnections(), hasKey("pcx-f754069e"));
+    assertThat(awsVpcConfiguration.getVpcPeeringConnections(), not(hasKey("pcx-4ee8b427")));
+  }
+}

--- a/projects/batfish/src/test/java/org/batfish/main/BatfishTestUtils.java
+++ b/projects/batfish/src/test/java/org/batfish/main/BatfishTestUtils.java
@@ -69,6 +69,49 @@ public class BatfishTestUtils {
     return batfish;
   }
 
+  public static Batfish getBatfishFromTestrigText(TestrigText input, TemporaryFolder tempFolder)
+      throws IOException {
+    SortedMap<String, String> awsText = input.getAwsText();
+    SortedMap<String, String> bgpTablesText = input.getBgpTablesText();
+    SortedMap<String, String> configurationText = input.getConfigurationText();
+    SortedMap<String, String> hostsText = input.getHostsText();
+    SortedMap<String, String> iptablesFilesText = input.getIptablesFilesText();
+    SortedMap<String, String> routingTablesText = input.getRoutingTablesText();
+
+    Settings settings = new Settings(new String[] {});
+    settings.setLogger(new BatfishLogger("debug", false));
+    settings.setDisableUnrecognized(true);
+    settings.setHaltOnConvertError(true);
+    settings.setHaltOnParseError(true);
+    settings.setThrowOnLexerError(true);
+    settings.setThrowOnParserError(true);
+    settings.setVerboseParse(true);
+    Path containerDir = tempFolder.newFolder("container").toPath();
+    settings.setContainerDir(containerDir);
+    settings.setTestrig("tempTestrig");
+    settings.setEnvironmentName("tempEnvironment");
+    Batfish.initTestrigSettings(settings);
+    Path testrigPath = settings.getBaseTestrigSettings().getTestRigPath();
+    settings.setActiveTestrigSettings(settings.getBaseTestrigSettings());
+    EnvironmentSettings envSettings = settings.getBaseTestrigSettings().getEnvironmentSettings();
+    envSettings.getEnvironmentBasePath().toFile().mkdirs();
+    writeTemporaryTestrigFiles(
+        configurationText, testrigPath.resolve(BfConsts.RELPATH_CONFIGURATIONS_DIR));
+    writeTemporaryTestrigFiles(awsText, testrigPath.resolve(BfConsts.RELPATH_AWS_VPC_CONFIGS_DIR));
+    writeTemporaryTestrigFiles(bgpTablesText, envSettings.getEnvironmentBgpTablesPath());
+    writeTemporaryTestrigFiles(hostsText, testrigPath.resolve(BfConsts.RELPATH_HOST_CONFIGS_DIR));
+    writeTemporaryTestrigFiles(iptablesFilesText, testrigPath.resolve("iptables"));
+    writeTemporaryTestrigFiles(routingTablesText, envSettings.getEnvironmentRoutingTablesPath());
+    Batfish batfish =
+        new Batfish(
+            settings,
+            makeTestrigCache(),
+            makeDataPlaneCache(),
+            makeEnvBgpCache(),
+            makeEnvRouteCache());
+    return batfish;
+  }
+
   private static Batfish initBatfishFromTestrigText(
       @Nullable SortedMap<String, String> configurationText,
       @Nullable SortedMap<String, String> bgpTablesText,

--- a/projects/batfish/src/test/java/org/batfish/main/TestrigText.java
+++ b/projects/batfish/src/test/java/org/batfish/main/TestrigText.java
@@ -1,0 +1,182 @@
+package org.batfish.main;
+
+import java.util.SortedMap;
+import java.util.SortedSet;
+import java.util.TreeMap;
+import org.batfish.common.BfConsts;
+import org.batfish.common.util.CommonUtil;
+
+public class TestrigText {
+
+  public static class Builder {
+
+    private SortedMap<String, String> _awsText;
+
+    private SortedMap<String, String> _bgpTablesText;
+
+    private SortedMap<String, String> _configurationText;
+
+    private SortedMap<String, String> _hostsText;
+
+    private SortedMap<String, String> _iptablesFilesText;
+
+    private SortedMap<String, String> _routingTablesText;
+
+    public TestrigText build() {
+      TestrigText testrigText = new TestrigText();
+      testrigText.setAwsText(_awsText);
+      testrigText.setBgpTablesText(_bgpTablesText);
+      testrigText.setConfigurationText(_configurationText);
+      testrigText.setHostsText(_hostsText);
+      testrigText.setIptablesFilesText(_iptablesFilesText);
+      testrigText.setRoutingTablesText(_routingTablesText);
+      return testrigText;
+    }
+
+    private static SortedMap<String, String> readTestrigResources(
+        String testrigResourcePrefix, String subfolder, SortedSet<String> filenames) {
+      SortedMap<String, String> content = new TreeMap<>();
+      if (filenames != null) {
+        for (String filename : filenames) {
+          String path = String.format("%s/%s/%s", testrigResourcePrefix, subfolder, filename);
+          String text = CommonUtil.readResource(path);
+          content.put(filename, text);
+        }
+      }
+      return content;
+    }
+
+    public Builder setAwsText(String testrigResourcePrefix, SortedSet<String> filenames) {
+      _awsText =
+          readTestrigResources(
+              testrigResourcePrefix, BfConsts.RELPATH_AWS_VPC_CONFIGS_DIR, filenames);
+      return this;
+    }
+
+    public Builder setAwsText(SortedMap<String, String> awsText) {
+      _awsText = awsText;
+      return this;
+    }
+
+    public Builder setBgpTablesText(String testrigResourcePrefix, SortedSet<String> filenames) {
+      _bgpTablesText =
+          readTestrigResources(
+              testrigResourcePrefix, BfConsts.RELPATH_ENVIRONMENT_BGP_TABLES, filenames);
+      return this;
+    }
+
+    public Builder setBgpTablesText(SortedMap<String, String> bgpTablesText) {
+      _bgpTablesText = bgpTablesText;
+      return this;
+    }
+
+    public Builder setConfigurationText(String testrigResourcePrefix, SortedSet<String> filenames) {
+      _configurationText =
+          readTestrigResources(
+              testrigResourcePrefix, BfConsts.RELPATH_CONFIGURATIONS_DIR, filenames);
+      return this;
+    }
+
+    public Builder setConfigurationText(SortedMap<String, String> configurationText) {
+      _configurationText = configurationText;
+      return this;
+    }
+
+    public Builder setHostsText(String testrigResourcePrefix, SortedSet<String> filenames) {
+      _hostsText =
+          readTestrigResources(testrigResourcePrefix, BfConsts.RELPATH_HOST_CONFIGS_DIR, filenames);
+      return this;
+    }
+
+    public Builder setHostsText(SortedMap<String, String> hostsText) {
+      _hostsText = hostsText;
+      return this;
+    }
+
+    public Builder setIptablesFilesText(String testrigResourcePrefix, SortedSet<String> filenames) {
+      _iptablesFilesText = readTestrigResources(testrigResourcePrefix, "iptables", filenames);
+      return this;
+    }
+
+    public Builder setIptablesFilesText(SortedMap<String, String> iptablesFilesText) {
+      _iptablesFilesText = iptablesFilesText;
+      return this;
+    }
+
+    public Builder setRoutingTablesText(String testrigResourcePrefix, SortedSet<String> filenames) {
+      _routingTablesText =
+          readTestrigResources(
+              testrigResourcePrefix, BfConsts.RELPATH_ENVIRONMENT_ROUTING_TABLES, filenames);
+      return this;
+    }
+
+    public Builder setRoutingTablesText(SortedMap<String, String> routingTablesText) {
+      _routingTablesText = routingTablesText;
+      return this;
+    }
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  private SortedMap<String, String> _awsText;
+
+  private SortedMap<String, String> _bgpTablesText;
+
+  private SortedMap<String, String> _configurationText;
+
+  private SortedMap<String, String> _hostsText;
+
+  private SortedMap<String, String> _iptablesFilesText;
+
+  private SortedMap<String, String> _routingTablesText;
+
+  public SortedMap<String, String> getAwsText() {
+    return _awsText;
+  }
+
+  public SortedMap<String, String> getBgpTablesText() {
+    return _bgpTablesText;
+  }
+
+  public SortedMap<String, String> getConfigurationText() {
+    return _configurationText;
+  }
+
+  public SortedMap<String, String> getHostsText() {
+    return _hostsText;
+  }
+
+  public SortedMap<String, String> getIptablesFilesText() {
+    return _iptablesFilesText;
+  }
+
+  public SortedMap<String, String> getRoutingTablesText() {
+    return _routingTablesText;
+  }
+
+  public void setAwsText(SortedMap<String, String> awsText) {
+    _awsText = awsText;
+  }
+
+  public void setBgpTablesText(SortedMap<String, String> bgpTablesText) {
+    _bgpTablesText = bgpTablesText;
+  }
+
+  public void setConfigurationText(SortedMap<String, String> configurationText) {
+    _configurationText = configurationText;
+  }
+
+  public void setHostsText(SortedMap<String, String> hostsText) {
+    _hostsText = hostsText;
+  }
+
+  public void setIptablesFilesText(SortedMap<String, String> iptablesFilesText) {
+    _iptablesFilesText = iptablesFilesText;
+  }
+
+  public void setRoutingTablesText(SortedMap<String, String> routingTablesText) {
+    _routingTablesText = routingTablesText;
+  }
+}

--- a/projects/batfish/src/test/resources/org/batfish/grammar/aws/testrigs/vpc_peering_connections/aws_configs/VpcPeeringConnections.json
+++ b/projects/batfish/src/test/resources/org/batfish/grammar/aws/testrigs/vpc_peering_connections/aws_configs/VpcPeeringConnections.json
@@ -1,0 +1,76 @@
+{
+ "VpcPeeringConnections": [
+  {
+   "AccepterVpcInfo": {
+    "CidrBlock": "10.199.100.0/24",
+    "CidrBlockSet": [
+     {
+      "CidrBlock": "10.199.100.0/24"
+     }
+    ],
+    "OwnerId": "118292266645",
+    "PeeringOptions": {
+     "AllowDnsResolutionFromRemoteVpc": false,
+     "AllowEgressFromLocalClassicLinkToRemoteVpc": false,
+     "AllowEgressFromLocalVpcToRemoteClassicLink": false
+    },
+    "VpcId": "vpc-f6c5c790"
+   },
+   "RequesterVpcInfo": {
+    "CidrBlock": "10.130.0.0/16",
+    "CidrBlockSet": [
+     {
+      "CidrBlock": "10.130.0.0/16"
+     }
+    ],
+    "OwnerId": "118292266645",
+    "PeeringOptions": {
+     "AllowDnsResolutionFromRemoteVpc": false,
+     "AllowEgressFromLocalClassicLinkToRemoteVpc": false,
+     "AllowEgressFromLocalVpcToRemoteClassicLink": false
+    },
+    "VpcId": "vpc-07acbc61"
+   },
+   "Status": {
+    "Code": "active",
+    "Message": "Active"
+   },
+   "Tags": [
+    {
+     "Key": "Name",
+     "Value": "peering-test"
+    }
+   ],
+   "VpcPeeringConnectionId": "pcx-f754069e"
+  },
+  {
+   "AccepterVpcInfo": {
+    "OwnerId": "118292266645",
+    "VpcId": "vpc-6ee90d17"
+   },
+   "RequesterVpcInfo": {
+    "OwnerId": "118292266645",
+    "VpcId": "vpc-56c5212f"
+   },
+   "Status": {
+    "Code": "deleted",
+    "Message": "Deleted by 118292266645"
+   },
+   "Tags": [
+    {
+     "Key": "aws:cloudformation:logical-id",
+     "Value": "VPCPeeringConnection"
+    },
+    {
+     "Key": "aws:cloudformation:stack-id",
+     "Value": "arn:aws:cloudformation:us-west-2:118292266645:stack/batfish-stack-test-1-VPC-1RWP34XRWCZQR/8ced1590-dad0-11e7-bf5d-50a68a2012ba"
+    },
+    {
+     "Key": "aws:cloudformation:stack-name",
+     "Value": "batfish-stack-test-1-VPC-1RWP34XRWCZQR"
+    }
+   ],
+   "VpcPeeringConnectionId": "pcx-4ee8b427"
+  }
+ ]
+}

